{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d56c51f1-21c1-4a13-a335-7156aa1638ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NC_000001.11\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/tmp/ipykernel_9485/1938491882.py:61: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_filtered['centred_sequence'] = df_filtered.apply(get_centered_sequence, axis=1)\n",
      "/var/tmp/ipykernel_9485/1938491882.py:77: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_filtered['centred_sequence_'] = df_filtered.apply(lambda row: reverse_complement(row['centred_sequence'].upper()) if row['strand'] == '-' else row['centred_sequence'], axis=1)\n",
      "/var/tmp/ipykernel_9485/1938491882.py:78: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_filtered.drop(columns = [\"centred_sequence\"], inplace = True)\n",
      "/var/tmp/ipykernel_9485/1938491882.py:79: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_filtered.rename(columns = {\"centred_sequence_\" : \"centred_sequence\"}, inplace = True)\n"
     ]
    }
   ],
   "source": [
    "sequence_length_left = 0\n",
    "sequence_length_right = 17500\n",
    "\n",
    "\n",
    "\n",
    "# Les deux listes\n",
    "liste_nc = [\n",
    "    'NC_000001.11', 'NC_000002.12', 'NC_000003.12', 'NC_000004.12', 'NC_000005.10',\n",
    "    'NC_000006.12', 'NC_000007.14', 'NC_000008.11', 'NC_000009.12',\n",
    "    'NC_000010.11', 'NC_000011.10', 'NC_000012.12', 'NC_000013.11',\n",
    "    'NC_000014.9', 'NC_000015.10', 'NC_000016.10', 'NC_000017.11',\n",
    "    'NC_000018.10', 'NC_000019.10', 'NC_000020.11', 'NC_000021.9',\n",
    "    'NC_000022.11', 'NC_000023.11', 'NC_000024.10'\n",
    "]\n",
    "\n",
    "liste_chr = [\n",
    "    'chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8',\n",
    "    'chr9', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16',\n",
    "    'chr17', 'chr18', 'chr19', 'chr20', 'chr21', 'chr22', 'chrX',\n",
    "    'chrY'\n",
    "]\n",
    "\n",
    "# Création du dictionnaire dans l'ordre demandé\n",
    "dex_to_ncbi = dict(zip(liste_chr, liste_nc))\n",
    "\n",
    "# Affichage du dictionnaire\n",
    "print(dex_to_ncbi['chr1'])\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Charger le fichier FASTA et stocker les séquences dans un dictionnaire\n",
    "fasta_file = \"../ncbi_dataset/data/GCF_000001405.40/GCF_000001405.40_GRCh38.p14_genomic.fna\"\n",
    "chromosome_sequences = {}\n",
    "for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "    chromosome_sequences[record.id] = str(record.seq)\n",
    "\n",
    "# Charger le DataFrame\n",
    "df = pd.read_csv(\"dex_annotations_important.csv\")\n",
    "df = df[(df['end'] - df['start']) <= 17_500]\n",
    "df_filtered = df[df['seqname'] != 'chrM']\n",
    "\n",
    "# Définir la longueur de la séquence à extraire (par exemple, 5000 bases de chaque côté)\n",
    "\n",
    "# Fonction pour extraire la séquence centrée autour de 'start' ou 'end' basé sur 'strand'\n",
    "def get_centered_sequence(row):\n",
    "    chromosome = dex_to_ncbi[row['seqname']]\n",
    "    if row['strand'] == '+':\n",
    "        center = row['start']\n",
    "    else:\n",
    "        center = row['end']\n",
    "    \n",
    "    seq = chromosome_sequences.get(chromosome, \"\")\n",
    "    if seq:\n",
    "        start_idx = max(0, center - sequence_length_left)\n",
    "        end_idx = min(len(seq), center + sequence_length_right)\n",
    "        return seq[start_idx-1:end_idx]\n",
    "    return \"\"\n",
    "\n",
    "# Ajouter la nouvelle colonne avec les séquences extraites\n",
    "df_filtered['centred_sequence'] = df_filtered.apply(get_centered_sequence, axis=1)\n",
    "\n",
    "def reverse_complement(sequence):\n",
    "    # Dictionnaire de complémentarité des nucléotides\n",
    "    complement = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C',\n",
    "                  'a': 't', 't': 'a', 'c': 'g', 'g': 'c',\n",
    "                  'N': 'N', 'n': 'n'}  # 'N' reste 'N', 'n' reste 'n'\n",
    "    \n",
    "    # Générer la séquence complémentaire\n",
    "    complement_sequence = ''.join(complement.get(base, base) for base in sequence)\n",
    "    \n",
    "    # Inverser la séquence complémentaire\n",
    "    reverse_complement_sequence = complement_sequence[::-1]\n",
    "    \n",
    "    return reverse_complement_sequence\n",
    "\n",
    "df_filtered['centred_sequence_'] = df_filtered.apply(lambda row: reverse_complement(row['centred_sequence'].upper()) if row['strand'] == '-' else row['centred_sequence'], axis=1)\n",
    "df_filtered.drop(columns = [\"centred_sequence\"], inplace = True)\n",
    "df_filtered.rename(columns = {\"centred_sequence_\" : \"centred_sequence\"}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "094c9ea9-af1a-4c65-9118-ee58bcd07246",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_filtered"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25c70c7b-6649-4383-b141-611901de6454",
   "metadata": {},
   "source": [
    "## Function to create the directory "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "cce5e931-118e-41e8-9042-ab004d792a4f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "\n",
    "def create_or_empty_directory(dir_path):\n",
    "    if os.path.exists(dir_path):\n",
    "        # If directory exists, delete all its contents\n",
    "        shutil.rmtree(dir_path)\n",
    "    # Create the directory (or recreate it if it was deleted)\n",
    "    os.makedirs(dir_path)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13ca2694-294f-484f-bc73-2a4cc7763a03",
   "metadata": {},
   "source": [
    "## Create the train test set similar as dexter "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d8ac462-7ccd-47f2-ae39-7d21a1d7ec36",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "path_ = \"DNABERT_2/DATASETS/GRCH38/less_than_17_500_tss\"\n",
    "\n",
    "create_or_empty_directory(path_)\n",
    "\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "# Load the initial DataFrame\n",
    "pituitary = pd.read_csv(\"human_genes_expression_pit/created_by_me/final_result.csv\")\n",
    "final_df = pd.merge(df_filtered, pituitary, left_on=\"GENE_ID\", right_on='gene_id')\n",
    "\n",
    "# Replace zeros with a small value to avoid errors in log transformation\n",
    "final_df['Pituitary'] = final_df['Pituitary'].replace(0, 0.001)\n",
    "\n",
    "# Take the logarithm base 10 of the 'Pituitary' column\n",
    "final_df['Pituitary'] = np.log10(final_df['Pituitary'])\n",
    "\n",
    "# Select desired columns\n",
    "final_df = final_df[['centred_sequence', 'Pituitary', 'last_gene_id', 'GENE_ID', 'dex_gene_id']]\n",
    "\n",
    "# Convert 'centred_sequence' to uppercase\n",
    "final_df['centred_sequence'] = final_df['centred_sequence'].str.upper()\n",
    "\n",
    "# Load the training set data\n",
    "train_dex = pd.read_csv(\"DExTER/example/my_experience_human/training_set.log\", names=[\"gene_id_dex_version\"])\n",
    "\n",
    "# Merge with final_df to get training data with required columns\n",
    "data_train_dex = pd.merge(train_dex, final_df, how=\"left\", left_on=\"gene_id_dex_version\", right_on=\"dex_gene_id\")\n",
    "\n",
    "# Select specific columns for training data\n",
    "data_train_dex = data_train_dex[['centred_sequence', 'Pituitary', 'last_gene_id']]\n",
    "\n",
    "# Remove rows with NaN values (if any)\n",
    "data_train_dex = data_train_dex.dropna()\n",
    "\n",
    "# Splitting into train (87%) and dev (13%)\n",
    "train_df, dev_df = train_test_split(data_train_dex, test_size=0.13, random_state=42)\n",
    "\n",
    "\n",
    "# Save train and dev datasets to CSV files\n",
    "train_df.to_csv(path_ + \"/train.csv\", index=False)\n",
    "dev_df.to_csv(path_ + \"/dev.csv\", index=False)\n",
    "\n",
    "# Load the testing set data\n",
    "test_dex = pd.read_csv(\"DExTER/example/my_experience_human/testing_set.log\", names=[\"gene_id_dex_version\"])\n",
    "\n",
    "# Merge with final_df to get testing data with required columns\n",
    "data_test_dex = pd.merge(test_dex, final_df, how=\"left\", left_on=\"gene_id_dex_version\", right_on=\"dex_gene_id\")\n",
    "\n",
    "# Select specific columns for testing data\n",
    "data_test_dex = data_test_dex[['centred_sequence', 'Pituitary', 'last_gene_id']]\n",
    "\n",
    "# Remove rows with NaN values (if any)\n",
    "data_test_dex = data_test_dex.dropna()\n",
    "\n",
    "# Save testing dataset to a CSV file\n",
    "data_test_dex.to_csv(path_ + \"/test.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59ed2e58-223d-480f-931c-f67736ea58fa",
   "metadata": {},
   "source": [
    "## less than"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b7f5a30d-fa14-4fa2-80cb-54c1cc5407e7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "path_ = \"DNABERT_2/DATASETS/GRCH38/less_than_17_500\"\n",
    "\n",
    "create_or_empty_directory(path_)\n",
    "\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "# Load the initial DataFrame\n",
    "pituitary = pd.read_csv(\"human_genes_expression_pit/created_by_me/final_result.csv\")\n",
    "final_df = pd.merge(df_filtered, pituitary, left_on=\"GENE_ID\", right_on='gene_id')\n",
    "\n",
    "# Replace zeros with a small value to avoid errors in log transformation\n",
    "final_df['Pituitary'] = final_df['Pituitary'].replace(0, 0.001)\n",
    "\n",
    "# Take the logarithm base 10 of the 'Pituitary' column\n",
    "final_df['Pituitary'] = np.log10(final_df['Pituitary'])\n",
    "\n",
    "# Select desired columns\n",
    "final_df = final_df[['centred_sequence', 'Pituitary', 'last_gene_id', 'GENE_ID', 'dex_gene_id']]\n",
    "\n",
    "# Convert 'centred_sequence' to uppercase\n",
    "final_df['centred_sequence'] = final_df['centred_sequence'].str.upper()\n",
    "\n",
    "\n",
    "# Splitting into train (87%) and dev (13%)\n",
    "train_df, test_df = train_test_split(final_df, test_size=0.20, random_state=42)\n",
    "test_df, dev_df = train_test_split(test_df, test_size =0.5, random_state=42)\n",
    "\n",
    "train_df = train_df[['centred_sequence', 'Pituitary', 'last_gene_id']]\n",
    "test_df = test_df[['centred_sequence', 'Pituitary', 'last_gene_id']]\n",
    "dev_df = dev_df[['centred_sequence', 'Pituitary', 'last_gene_id']]\n",
    "\n",
    "\n",
    "\n",
    "# Save train and dev datasets to CSV files\n",
    "train_df.to_csv(path_ + \"/train.csv\", index=False)\n",
    "test_df.to_csv(path_ + \"/test.csv\", index=False)\n",
    "dev_df.to_csv(path_ + \"/dev.csv\", index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "4f76ca8f-84f6-4d49-bc4c-def54f746833",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>centred_sequence</th>\n",
       "      <th>Pituitary</th>\n",
       "      <th>last_gene_id</th>\n",
       "      <th>GENE_ID</th>\n",
       "      <th>dex_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4716</th>\n",
       "      <td>AGACTTCCTGCCCCTGCTCTGCACTCTCAGGTAGGGACCTGGCTCT...</td>\n",
       "      <td>-0.589341</td>\n",
       "      <td>ENSG00000233041.8</td>\n",
       "      <td>ENSG00000233041</td>\n",
       "      <td>ENSG00000233041.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1272</th>\n",
       "      <td>CCTTCAGACGAAATCCTTGGGACAGGCAAAACCAGGGGCAGGATGT...</td>\n",
       "      <td>-1.093001</td>\n",
       "      <td>ENSG00000240541.3</td>\n",
       "      <td>ENSG00000240541</td>\n",
       "      <td>ENSG00000240541.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6212</th>\n",
       "      <td>CTCATTAACTCCATCTTCAGTGATTTGGCCAGAACCTGCTGCCAGT...</td>\n",
       "      <td>-0.396343</td>\n",
       "      <td>ENSG00000215595.3</td>\n",
       "      <td>ENSG00000215595</td>\n",
       "      <td>ENSG00000215595.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4487</th>\n",
       "      <td>TCGGTCACAATTTGACATATTGTTATTTAAATTTTAATGTAAACCT...</td>\n",
       "      <td>1.368729</td>\n",
       "      <td>ENSG00000258441.1</td>\n",
       "      <td>ENSG00000258441</td>\n",
       "      <td>ENSG00000258441.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6052</th>\n",
       "      <td>AAAATTAGCCGGGCGTGGAGGCTTAGGCAGGAGGATGGCGTGAACC...</td>\n",
       "      <td>1.930117</td>\n",
       "      <td>ENSG00000126456.16</td>\n",
       "      <td>ENSG00000126456</td>\n",
       "      <td>ENSG00000126456.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3772</th>\n",
       "      <td>ACCGGCTGAGCCATGGATACACCGTTAAGGCGCAGCCGACGGCTGG...</td>\n",
       "      <td>1.578425</td>\n",
       "      <td>ENSG00000110104.13</td>\n",
       "      <td>ENSG00000110104</td>\n",
       "      <td>ENSG00000110104.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5191</th>\n",
       "      <td>TCAAGAAAGGAGGAGGTGGGGGATCCTCAGAGGGAATAAAAAAGGT...</td>\n",
       "      <td>1.702449</td>\n",
       "      <td>ENSG00000004975.12</td>\n",
       "      <td>ENSG00000004975</td>\n",
       "      <td>ENSG00000004975.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5226</th>\n",
       "      <td>GCAGATTAGAACATGATGATTGGAGATGCATGAAACGTGATTAACG...</td>\n",
       "      <td>2.454645</td>\n",
       "      <td>ENSG00000220205.10</td>\n",
       "      <td>ENSG00000220205</td>\n",
       "      <td>ENSG00000220205.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5390</th>\n",
       "      <td>AAAAGTTATTGCGGTTTTGGCTGCCAAAAGAAATGGCCAAAACCGC...</td>\n",
       "      <td>-3.000000</td>\n",
       "      <td>ENSG00000264314.1</td>\n",
       "      <td>ENSG00000264314</td>\n",
       "      <td>ENSG00000264314.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>AGTCGAGGTTGGCTTACCAGCTTCTTCCTAGAAGGGAGAAATACCT...</td>\n",
       "      <td>-0.994271</td>\n",
       "      <td>ENSG00000230140.7</td>\n",
       "      <td>ENSG00000230140</td>\n",
       "      <td>ENSG00000230140.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5310 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       centred_sequence  Pituitary  \\\n",
       "4716  AGACTTCCTGCCCCTGCTCTGCACTCTCAGGTAGGGACCTGGCTCT...  -0.589341   \n",
       "1272  CCTTCAGACGAAATCCTTGGGACAGGCAAAACCAGGGGCAGGATGT...  -1.093001   \n",
       "6212  CTCATTAACTCCATCTTCAGTGATTTGGCCAGAACCTGCTGCCAGT...  -0.396343   \n",
       "4487  TCGGTCACAATTTGACATATTGTTATTTAAATTTTAATGTAAACCT...   1.368729   \n",
       "6052  AAAATTAGCCGGGCGTGGAGGCTTAGGCAGGAGGATGGCGTGAACC...   1.930117   \n",
       "...                                                 ...        ...   \n",
       "3772  ACCGGCTGAGCCATGGATACACCGTTAAGGCGCAGCCGACGGCTGG...   1.578425   \n",
       "5191  TCAAGAAAGGAGGAGGTGGGGGATCCTCAGAGGGAATAAAAAAGGT...   1.702449   \n",
       "5226  GCAGATTAGAACATGATGATTGGAGATGCATGAAACGTGATTAACG...   2.454645   \n",
       "5390  AAAAGTTATTGCGGTTTTGGCTGCCAAAAGAAATGGCCAAAACCGC...  -3.000000   \n",
       "860   AGTCGAGGTTGGCTTACCAGCTTCTTCCTAGAAGGGAGAAATACCT...  -0.994271   \n",
       "\n",
       "            last_gene_id          GENE_ID         dex_gene_id  \n",
       "4716   ENSG00000233041.8  ENSG00000233041   ENSG00000233041.4  \n",
       "1272   ENSG00000240541.3  ENSG00000240541   ENSG00000240541.2  \n",
       "6212   ENSG00000215595.3  ENSG00000215595   ENSG00000215595.1  \n",
       "4487   ENSG00000258441.1  ENSG00000258441   ENSG00000258441.1  \n",
       "6052  ENSG00000126456.16  ENSG00000126456  ENSG00000126456.11  \n",
       "...                  ...              ...                 ...  \n",
       "3772  ENSG00000110104.13  ENSG00000110104   ENSG00000110104.7  \n",
       "5191  ENSG00000004975.12  ENSG00000004975   ENSG00000004975.7  \n",
       "5226  ENSG00000220205.10  ENSG00000220205   ENSG00000220205.4  \n",
       "5390   ENSG00000264314.1  ENSG00000264314   ENSG00000264314.1  \n",
       "860    ENSG00000230140.7  ENSG00000230140   ENSG00000230140.1  \n",
       "\n",
       "[5310 rows x 5 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d12a9ac-def3-4fa2-9da6-34ac402d88d0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "environment": {
   "kernel": "conda-base-py",
   "name": "workbench-notebooks.m123",
   "type": "gcloud",
   "uri": "us-docker.pkg.dev/deeplearning-platform-release/gcr.io/workbench-notebooks:m123"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
