{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8e25f165-1de2-41dd-b2a0-c07c7170ab61",
   "metadata": {},
   "source": [
    "# Retrieve genes sequence around the TSS (transcript start site) with its median gene expression "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83ade290-0850-4b90-8e77-d8cf94a51491",
   "metadata": {
    "tags": []
   },
   "source": [
    "This notebook is designed to retrieve gene sequences and their corresponding median expression levels. To make it work, you'll need to adjust the file paths for the following three required files:\n",
    "\n",
    "1. **Gene Expression Data**: A file from GTEx containing median gene expression levels.\n",
    "2. **Human Genome Sequence**: A file from NCBI containing the human genome sequence.\n",
    "3. **Genome Annotations**: A file from NCBI with annotations for the human genome.\n",
    "\n",
    "Once you've set the correct paths, the notebook will be ready to use. Ctrl+f \"path_to_change\" to find the paths\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a427a2b5-0932-46d0-bd9c-ec070fb14b0e",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Genes sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da35743c-b4f7-4f25-9082-fb548f9ec464",
   "metadata": {},
   "source": [
    "To begin, we should have the IDs of the genes for which we want to retrieve sequences. For this, we need to download the complete human genome from NCBI. Here is the link for GRCh38.p14: [GRCh38.p14 Genome](https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000001405.40/). Make sure that there isn't a more recent version of the human genome available, as it is constantly being updated and corrected.  \n",
    "<br>\n",
    "Download **GenBank** format only, and ensure that you select **genome sequences** as well as **sequence and annotations**.  \n",
    "<br>\n",
    "Once it's downloaded, unzip the file. You'll need two files: one that contains the gene sequences, and the other that contains the annotations. To retrieve the sequence of a specific gene, you'll need information on its start, end, strand, chromosome and gene ID.  \n",
    "<br>\n",
    "The file containing the annotations is named `genomic.gff`, and the one containing the gene sequences is named `GCF_000001405.40_GRCh38.p14_genomic.fna`. <br>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a073129a-6a33-4d37-ad33-c851ee08b15d",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Annotations table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76d238e2-4e32-4d83-b69b-88de083c78b8",
   "metadata": {},
   "source": [
    "You can see in the following the columns that are in the annotations table for all the genes you can filter the dataframe based on your preferences for example if you want only protein_coding genes you filter your dataframe based on the column **gene_biotype**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d2a283d-595d-40a8-92c2-5814953f486e",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install gtfparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "34c71e6e-fa63-4b1b-929f-abf4429cf12d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'transcript_id', 'db_xref', 'description', 'gbkey', 'gene', 'gene_biotype', 'pseudo', 'product', 'transcript_biotype', 'exon_number', 'gene_synonym', 'model_evidence', 'tag', 'protein_id', 'experiment', 'inference', 'note', 'part', 'exception', 'isoform', 'anticodon', 'partial', 'The', 'transl_except', 'non-AUG', 'standard_name', 'deleted', 'source', 'similar', 'substituted', 'transferase', 'codons', '12S', '16S', 'transl_table', 'ATPase']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_annotation columns : \n",
      "\n",
      "Index(['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand',\n",
      "       'frame', 'last_gene_id', 'transcript_id', 'db_xref', 'description',\n",
      "       'gbkey', 'gene', 'gene_biotype', 'pseudo', 'product',\n",
      "       'transcript_biotype', 'exon_number', 'gene_synonym', 'model_evidence',\n",
      "       'tag', 'protein_id', 'experiment', 'inference', 'note', 'part',\n",
      "       'exception', 'isoform', 'anticodon', 'partial', 'The', 'transl_except',\n",
      "       'non-AUG', 'standard_name', 'deleted', 'similar', 'substituted',\n",
      "       'transferase', 'codons', '12S', '16S', 'transl_table', 'ATPase',\n",
      "       'GENE_ID'],\n",
      "      dtype='object')\n",
      "\n",
      "\n",
      "example of the annotations of a gene : \n",
      "\n",
      "seqname                                   NC_000001.11\n",
      "source                                                \n",
      "feature                                           gene\n",
      "start                                            11874\n",
      "end                                              14409\n",
      "score                                              NaN\n",
      "strand                                               +\n",
      "frame                                                0\n",
      "last_gene_id                                   DDX11L1\n",
      "transcript_id                                         \n",
      "db_xref               GeneID:100287102,HGNC:HGNC:37102\n",
      "description                                 DEAD/H-box\n",
      "gbkey                                             Gene\n",
      "gene                                           DDX11L1\n",
      "gene_biotype                    transcribed_pseudogene\n",
      "pseudo                                            true\n",
      "product                                               \n",
      "transcript_biotype                                    \n",
      "exon_number                                           \n",
      "gene_synonym                                          \n",
      "model_evidence                                        \n",
      "tag                                                   \n",
      "protein_id                                            \n",
      "experiment                                            \n",
      "inference                                             \n",
      "note                                                  \n",
      "part                                                  \n",
      "exception                                             \n",
      "isoform                                               \n",
      "anticodon                                             \n",
      "partial                                               \n",
      "The                                                   \n",
      "transl_except                                         \n",
      "non-AUG                                               \n",
      "standard_name                                         \n",
      "deleted                                               \n",
      "similar                                               \n",
      "substituted                                           \n",
      "transferase                                           \n",
      "codons                                                \n",
      "12S                                                   \n",
      "16S                                                   \n",
      "transl_table                                          \n",
      "ATPase                                                \n",
      "GENE_ID                                        DDX11L1\n",
      "Name: 0, dtype: object\n"
     ]
    }
   ],
   "source": [
    "from gtfparse import read_gtf\n",
    "\n",
    "# returns GTF with essential columns such as \"feature\", \"seqname\", \"start\", \"end\"\n",
    "# alongside the names of any optional keys which appeared in the attribute column path_to_change\n",
    "gene_annotations = read_gtf(\"../ncbi_dataset/data/GCF_000001405.40/genomic.gtf\")\n",
    "\n",
    "gene_annotations = gene_annotations.to_pandas()\n",
    "\n",
    "gene_annotations = gene_annotations[gene_annotations[\"feature\"]==\"gene\"]\n",
    "\n",
    "gene_annotations[\"GENE_ID\"] = gene_annotations[\"gene_id\"].apply(lambda x: x.split(\".\")[0])\n",
    "gene_annotations.rename(columns = {\"gene_id\" : \"last_gene_id\"}, inplace = True)\n",
    "\n",
    "print(\"gene_annotation columns : \\n\")\n",
    "print(gene_annotations.columns)\n",
    "print(\"\\n\")\n",
    "print(\"example of the annotations of a gene : \\n\")\n",
    "print(gene_annotations.iloc[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82bc1def-93eb-427e-9642-8553963444c3",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Retrieve the Sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2beff4ad-5681-4072-85bd-7a0e0de4dc0a",
   "metadata": {},
   "source": [
    "Change here these two top varibales to retrieve the frame sequence of the gene around the TSS as you prefere"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b1c7830a-33f8-4ab2-8abd-56ce81c513db",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sequence_length_left = 0\n",
    "sequence_length_right = 2000\n",
    "\n",
    "\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Charger le fichier FASTA et stocker les séquences dans un dictionnaire path_to_change\n",
    "fasta_file = \"../ncbi_dataset/data/GCF_000001405.40/GCF_000001405.40_GRCh38.p14_genomic.fna\"\n",
    "chromosome_sequences = {}\n",
    "for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "    chromosome_sequences[record.id] = str(record.seq)\n",
    "\n",
    "# Charger le DataFrame\n",
    "df = gene_annotations\n",
    "\n",
    "# Définir la longueur de la séquence à extraire (par exemple, 5000 bases de chaque côté)\n",
    "\n",
    "# Fonction pour extraire la séquence centrée autour de 'start' ou 'end' basé sur 'strand'\n",
    "def get_centered_sequence(row):\n",
    "    chromosome = row['seqname']\n",
    "    if row['strand'] == '+':\n",
    "        center = row['start']\n",
    "    else:\n",
    "        center = row['end']\n",
    "    \n",
    "    seq = chromosome_sequences.get(chromosome, \"\")\n",
    "    if seq:\n",
    "        start_idx = max(0, center - sequence_length_left)\n",
    "        end_idx = min(len(seq), center + sequence_length_right)\n",
    "        return seq[start_idx-1:end_idx]\n",
    "    return \"\"\n",
    "\n",
    "# Ajouter la nouvelle colonne avec les séquences extraites\n",
    "df['sequence'] = df.apply(get_centered_sequence, axis=1)\n",
    "\n",
    "def reverse_complement(sequence):\n",
    "    # Dictionnaire de complémentarité des nucléotides\n",
    "    complement = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C',\n",
    "                  'a': 't', 't': 'a', 'c': 'g', 'g': 'c',\n",
    "                  'N': 'N', 'n': 'n'}  # 'N' reste 'N', 'n' reste 'n'\n",
    "    \n",
    "    # Générer la séquence complémentaire\n",
    "    complement_sequence = ''.join(complement.get(base, base) for base in sequence)\n",
    "    \n",
    "    # Inverser la séquence complémentaire\n",
    "    reverse_complement_sequence = complement_sequence[::-1]\n",
    "    \n",
    "    return reverse_complement_sequence\n",
    "\n",
    "df['sequence_'] = df.apply(lambda row: reverse_complement(row['sequence'].upper()) if row['strand'] == '-' else row['sequence'], axis=1)\n",
    "df.drop(columns = [\"sequence\"], inplace = True)\n",
    "df.rename(columns = {\"sequence_\" : \"sequence\"}, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d6bdaacd-99e1-4993-b506-85b3e7de1f39",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand',\n",
       "       'frame', 'last_gene_id', 'transcript_id', 'db_xref', 'description',\n",
       "       'gbkey', 'gene', 'gene_biotype', 'pseudo', 'product',\n",
       "       'transcript_biotype', 'exon_number', 'gene_synonym', 'model_evidence',\n",
       "       'tag', 'protein_id', 'experiment', 'inference', 'note', 'part',\n",
       "       'exception', 'isoform', 'anticodon', 'partial', 'The', 'transl_except',\n",
       "       'non-AUG', 'standard_name', 'deleted', 'similar', 'substituted',\n",
       "       'transferase', 'codons', '12S', '16S', 'transl_table', 'ATPase',\n",
       "       'GENE_ID', 'sequence'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53e87029-dee5-4452-a1a2-9f7c3f05f358",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Result"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2594c71b-0fc3-4f6f-b8e8-a53aeddc4fd4",
   "metadata": {
    "tags": []
   },
   "source": [
    "we have now our dataframe with the genes and their corresponding sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4e5c6f92-ec91-4593-a195-5004bbdc6f29",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>last_gene_id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>gene_biotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "      <td>transcribed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>14362</td>\n",
       "      <td>29370</td>\n",
       "      <td>-</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>AAATCCCATTACAAATGGGGTGACTGAAGCTCCATTCATGGCTTGC...</td>\n",
       "      <td>transcribed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>-</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>ACCATCCCACATCCTCCCCCTCCCTTTCCTCATGAACCCCAGTCGC...</td>\n",
       "      <td>miRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>29774</td>\n",
       "      <td>35418</td>\n",
       "      <td>+</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>tgccctccagccctaCGCCTTGACCCGCTTTCCTGCGTCTCTCAGC...</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>+</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>GGATGCCCAGCTagtttgaattttagataaacaacgAATAATTTCG...</td>\n",
       "      <td>miRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4693791</th>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>-</td>\n",
       "      <td>ND6</td>\n",
       "      <td>CATCGTGATGTCTTATTTAAGGGGAACGTGTGGGCTATTTAGGCTT...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4693797</th>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>-</td>\n",
       "      <td>TRNE</td>\n",
       "      <td>CATCGTGATGTCTTATTTAAGGGGAACGTGTGGGCTATTTAGGCTT...</td>\n",
       "      <td>tRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4693800</th>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>+</td>\n",
       "      <td>CYTB</td>\n",
       "      <td>ATGACCCCAATACGCAAAACTAACCCCCTAATAAAATTAATTAACC...</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4693806</th>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>+</td>\n",
       "      <td>TRNT</td>\n",
       "      <td>GTCCTTGTAGTATAAACTAATACACCAGTCTTGTAAACCGGAGATG...</td>\n",
       "      <td>tRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4693809</th>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>-</td>\n",
       "      <td>TRNP</td>\n",
       "      <td>CATCGTGATGTCTTATTTAAGGGGAACGTGTGGGCTATTTAGGCTT...</td>\n",
       "      <td>tRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67356 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              seqname  start    end strand last_gene_id  \\\n",
       "0        NC_000001.11  11874  14409      +      DDX11L1   \n",
       "5        NC_000001.11  14362  29370      -       WASH7P   \n",
       "18       NC_000001.11  17369  17436      -    MIR6859-1   \n",
       "25       NC_000001.11  29774  35418      +  MIR1302-2HG   \n",
       "30       NC_000001.11  30366  30503      +    MIR1302-2   \n",
       "...               ...    ...    ...    ...          ...   \n",
       "4693791   NC_012920.1  14149  14673      -          ND6   \n",
       "4693797   NC_012920.1  14674  14742      -         TRNE   \n",
       "4693800   NC_012920.1  14747  15887      +         CYTB   \n",
       "4693806   NC_012920.1  15888  15953      +         TRNT   \n",
       "4693809   NC_012920.1  15956  16023      -         TRNP   \n",
       "\n",
       "                                                  sequence  \\\n",
       "0        CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...   \n",
       "5        AAATCCCATTACAAATGGGGTGACTGAAGCTCCATTCATGGCTTGC...   \n",
       "18       ACCATCCCACATCCTCCCCCTCCCTTTCCTCATGAACCCCAGTCGC...   \n",
       "25       tgccctccagccctaCGCCTTGACCCGCTTTCCTGCGTCTCTCAGC...   \n",
       "30       GGATGCCCAGCTagtttgaattttagataaacaacgAATAATTTCG...   \n",
       "...                                                    ...   \n",
       "4693791  CATCGTGATGTCTTATTTAAGGGGAACGTGTGGGCTATTTAGGCTT...   \n",
       "4693797  CATCGTGATGTCTTATTTAAGGGGAACGTGTGGGCTATTTAGGCTT...   \n",
       "4693800  ATGACCCCAATACGCAAAACTAACCCCCTAATAAAATTAATTAACC...   \n",
       "4693806  GTCCTTGTAGTATAAACTAATACACCAGTCTTGTAAACCGGAGATG...   \n",
       "4693809  CATCGTGATGTCTTATTTAAGGGGAACGTGTGGGCTATTTAGGCTT...   \n",
       "\n",
       "                   gene_biotype  \n",
       "0        transcribed_pseudogene  \n",
       "5        transcribed_pseudogene  \n",
       "18                        miRNA  \n",
       "25                       lncRNA  \n",
       "30                        miRNA  \n",
       "...                         ...  \n",
       "4693791          protein_coding  \n",
       "4693797                    tRNA  \n",
       "4693800          protein_coding  \n",
       "4693806                    tRNA  \n",
       "4693809                    tRNA  \n",
       "\n",
       "[67356 rows x 7 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['seqname', 'start', 'end', 'strand', 'last_gene_id', 'sequence', 'gene_biotype']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98ff71de-506b-489f-979b-01f7f045dcc5",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Gene expression data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a16d7bac-5bdb-462d-95ae-428f544a09a4",
   "metadata": {
    "tags": []
   },
   "source": [
    "Now that we have our gene sequences, we need to obtain gene expression data to train a model that learns to associate gene sequences with their corresponding expressions.\n",
    "\n",
    "For each gene, the expression data can vary based on the individual and the tissue where the measurement is taken. To simplify this, we aggregate all measurements to produce a single value for each gene per tissue. In this example, we'll use data from the pituitary tissue, but you can choose any tissue you prefer.\n",
    "\n",
    "We retrieve the data from GTEx. Below is the link to a file that contains the median expression for each gene across various tissues:\n",
    "\n",
    "- [GTEx Download Link](https://gtexportal.org/home/downloads/adult-gtex/bulk_tissue_expression)\n",
    "\n",
    "Download the file named: **`GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct.gz`**. This file contains the median gene-level TPM (Transcripts Per Million) by tissue. The median expression values were calculated from the file `GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_tpm.gct.gz`. The file size is 6.6 MB.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "61196aa1-cf16-4669-a9ca-b23dadd0c2d3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Pituitary</th>\n",
       "      <th>Description</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>5.425460</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>ENSG00000278267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>ENSG00000237613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56195</th>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>2028.050000</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>ENSG00000198695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56196</th>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>3.642400</td>\n",
       "      <td>MT-TE</td>\n",
       "      <td>ENSG00000210194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56197</th>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>19925.500000</td>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>ENSG00000198727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56198</th>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>MT-TT</td>\n",
       "      <td>ENSG00000210195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56199</th>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>0.575686</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>ENSG00000210196</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56156 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Name     Pituitary  Description          gene_id\n",
       "0      ENSG00000223972.5      0.000000      DDX11L1  ENSG00000223972\n",
       "1      ENSG00000227232.5      5.425460       WASH7P  ENSG00000227232\n",
       "2      ENSG00000278267.1      0.000000    MIR6859-1  ENSG00000278267\n",
       "3      ENSG00000243485.5      0.000000  MIR1302-2HG  ENSG00000243485\n",
       "4      ENSG00000237613.2      0.000000      FAM138A  ENSG00000237613\n",
       "...                  ...           ...          ...              ...\n",
       "56195  ENSG00000198695.2   2028.050000       MT-ND6  ENSG00000198695\n",
       "56196  ENSG00000210194.1      3.642400        MT-TE  ENSG00000210194\n",
       "56197  ENSG00000198727.2  19925.500000       MT-CYB  ENSG00000198727\n",
       "56198  ENSG00000210195.2      0.000000        MT-TT  ENSG00000210195\n",
       "56199  ENSG00000210196.2      0.575686        MT-TP  ENSG00000210196\n",
       "\n",
       "[56156 rows x 4 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd \n",
    "med_exp = pd.read_csv('human_genes_expression_pit/created_by_me/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct', sep='\\t', skiprows= 2) # path_to_change\n",
    "med_exp = med_exp[['Name', 'Pituitary', 'Description']]\n",
    "med_exp['gene_id'] = med_exp['Name'].apply(lambda x: x.split('.')[0])\n",
    "med_exp = med_exp.drop_duplicates(subset=['gene_id'])\n",
    "med_exp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "306ab473-914e-4163-8d9a-3881c63236b7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>last_gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>...</th>\n",
       "      <th>12S</th>\n",
       "      <th>16S</th>\n",
       "      <th>transl_table</th>\n",
       "      <th>ATPase</th>\n",
       "      <th>GENE_ID</th>\n",
       "      <th>sequence</th>\n",
       "      <th>Name</th>\n",
       "      <th>Pituitary</th>\n",
       "      <th>Description</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td></td>\n",
       "      <td>gene</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td></td>\n",
       "      <td>gene</td>\n",
       "      <td>14362</td>\n",
       "      <td>29370</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>AAATCCCATTACAAATGGGGTGACTGAAGCTCCATTCATGGCTTGC...</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>5.42546</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td></td>\n",
       "      <td>gene</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>ACCATCCCACATCCTCCCCCTCCCTTTCCTCATGAACCCCAGTCGC...</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>ENSG00000278267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td></td>\n",
       "      <td>gene</td>\n",
       "      <td>29774</td>\n",
       "      <td>35418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>tgccctccagccctaCGCCTTGACCCGCTTTCCTGCGTCTCTCAGC...</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td></td>\n",
       "      <td>gene</td>\n",
       "      <td>34611</td>\n",
       "      <td>36081</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>CTCCTGTCAAACCCACGTGGAAGGCAGGCTCTGGGCTGTGTTACTG...</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>ENSG00000237613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34682</th>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td></td>\n",
       "      <td>gene</td>\n",
       "      <td>26563992</td>\n",
       "      <td>26579868</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>TPTE2P4</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>TPTE2P4</td>\n",
       "      <td>tttcttttttactgttatccaaaaaaattataaaagacatGAATTG...</td>\n",
       "      <td>ENSG00000215506.5</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>TPTE2P4</td>\n",
       "      <td>ENSG00000215506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34683</th>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td></td>\n",
       "      <td>gene</td>\n",
       "      <td>26586445</td>\n",
       "      <td>26587563</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>SLC25A15P1</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>SLC25A15P1</td>\n",
       "      <td>AAGGCAGAGGTTGAAGTGAGTGAAGATTGTGCCATTACACTCCAGC...</td>\n",
       "      <td>ENSG00000227629.1</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>SLC25A15P1</td>\n",
       "      <td>ENSG00000227629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34684</th>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td></td>\n",
       "      <td>gene</td>\n",
       "      <td>26594668</td>\n",
       "      <td>26634655</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>PARP4P1</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>PARP4P1</td>\n",
       "      <td>AAGCAAAGGATCTGGAATTTGTAAGGCTTAGCTTTGAATTCTTGCT...</td>\n",
       "      <td>ENSG00000237917.1</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>PARP4P1</td>\n",
       "      <td>ENSG00000237917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34685</th>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td></td>\n",
       "      <td>gene</td>\n",
       "      <td>56855243</td>\n",
       "      <td>56856370</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>CTBP2P1</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>CTBP2P1</td>\n",
       "      <td>aagaaactgtgtgAACAAGGAATTCTTTGTCACATCTGTGCTTTGG...</td>\n",
       "      <td>ENSG00000235857.1</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>CTBP2P1</td>\n",
       "      <td>ENSG00000235857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34686</th>\n",
       "      <td>NT_187562.1</td>\n",
       "      <td></td>\n",
       "      <td>gene</td>\n",
       "      <td>311143</td>\n",
       "      <td>311575</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>TRBV6-2</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>TRBV6-2</td>\n",
       "      <td>ATGAGCCTCGGGCTCCTGTGCTGTGGGGCCTTTTCTCTCCTGTGGG...</td>\n",
       "      <td>ENSG00000283063.1</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>TRBV6-2</td>\n",
       "      <td>ENSG00000283063</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34687 rows × 50 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            seqname source feature     start       end  score strand  frame  \\\n",
       "0      NC_000001.11           gene     11874     14409    NaN      +      0   \n",
       "1      NC_000001.11           gene     14362     29370    NaN      -      0   \n",
       "2      NC_000001.11           gene     17369     17436    NaN      -      0   \n",
       "3      NC_000001.11           gene     29774     35418    NaN      +      0   \n",
       "4      NC_000001.11           gene     34611     36081    NaN      -      0   \n",
       "...             ...    ...     ...       ...       ...    ...    ...    ...   \n",
       "34682  NC_000024.10           gene  26563992  26579868    NaN      +      0   \n",
       "34683  NC_000024.10           gene  26586445  26587563    NaN      -      0   \n",
       "34684  NC_000024.10           gene  26594668  26634655    NaN      -      0   \n",
       "34685  NC_000024.10           gene  56855243  56856370    NaN      +      0   \n",
       "34686   NT_187562.1           gene    311143    311575    NaN      +      0   \n",
       "\n",
       "      last_gene_id transcript_id  ... 12S 16S transl_table ATPase  \\\n",
       "0          DDX11L1                ...                               \n",
       "1           WASH7P                ...                               \n",
       "2        MIR6859-1                ...                               \n",
       "3      MIR1302-2HG                ...                               \n",
       "4          FAM138A                ...                               \n",
       "...            ...           ...  ...  ..  ..          ...    ...   \n",
       "34682      TPTE2P4                ...                               \n",
       "34683   SLC25A15P1                ...                               \n",
       "34684      PARP4P1                ...                               \n",
       "34685      CTBP2P1                ...                               \n",
       "34686      TRBV6-2                ...                               \n",
       "\n",
       "           GENE_ID                                           sequence  \\\n",
       "0          DDX11L1  CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...   \n",
       "1           WASH7P  AAATCCCATTACAAATGGGGTGACTGAAGCTCCATTCATGGCTTGC...   \n",
       "2        MIR6859-1  ACCATCCCACATCCTCCCCCTCCCTTTCCTCATGAACCCCAGTCGC...   \n",
       "3      MIR1302-2HG  tgccctccagccctaCGCCTTGACCCGCTTTCCTGCGTCTCTCAGC...   \n",
       "4          FAM138A  CTCCTGTCAAACCCACGTGGAAGGCAGGCTCTGGGCTGTGTTACTG...   \n",
       "...            ...                                                ...   \n",
       "34682      TPTE2P4  tttcttttttactgttatccaaaaaaattataaaagacatGAATTG...   \n",
       "34683   SLC25A15P1  AAGGCAGAGGTTGAAGTGAGTGAAGATTGTGCCATTACACTCCAGC...   \n",
       "34684      PARP4P1  AAGCAAAGGATCTGGAATTTGTAAGGCTTAGCTTTGAATTCTTGCT...   \n",
       "34685      CTBP2P1  aagaaactgtgtgAACAAGGAATTCTTTGTCACATCTGTGCTTTGG...   \n",
       "34686      TRBV6-2  ATGAGCCTCGGGCTCCTGTGCTGTGGGGCCTTTTCTCTCCTGTGGG...   \n",
       "\n",
       "                    Name Pituitary  Description          gene_id  \n",
       "0      ENSG00000223972.5   0.00000      DDX11L1  ENSG00000223972  \n",
       "1      ENSG00000227232.5   5.42546       WASH7P  ENSG00000227232  \n",
       "2      ENSG00000278267.1   0.00000    MIR6859-1  ENSG00000278267  \n",
       "3      ENSG00000243485.5   0.00000  MIR1302-2HG  ENSG00000243485  \n",
       "4      ENSG00000237613.2   0.00000      FAM138A  ENSG00000237613  \n",
       "...                  ...       ...          ...              ...  \n",
       "34682  ENSG00000215506.5   0.00000      TPTE2P4  ENSG00000215506  \n",
       "34683  ENSG00000227629.1   0.00000   SLC25A15P1  ENSG00000227629  \n",
       "34684  ENSG00000237917.1   0.00000      PARP4P1  ENSG00000237917  \n",
       "34685  ENSG00000235857.1   0.00000      CTBP2P1  ENSG00000235857  \n",
       "34686  ENSG00000283063.1   0.00000      TRBV6-2  ENSG00000283063  \n",
       "\n",
       "[34687 rows x 50 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df = pd.merge(df, med_exp, left_on='last_gene_id', right_on='Description', how='inner')\n",
    "final_df"
   ]
  }
 ],
 "metadata": {
  "environment": {
   "kernel": "conda-base-py",
   "name": "workbench-notebooks.m123",
   "type": "gcloud",
   "uri": "us-docker.pkg.dev/deeplearning-platform-release/gcr.io/workbench-notebooks:m123"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
